src/lib.rs

```diff
@@ -1034,20 +1035,80 @@ impl<T: Clone + Integer + Signed> Signed for Ratio<T> {
 }
 
 // String conversions
-impl<T> fmt::Display for Ratio<T>
-where
-    T: fmt::Display + Eq + One,
-{
-    /// Renders as `numer/denom`. If denom=1, renders as numer.
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        if self.denom.is_one() {
-            write!(f, "{}", self.numer)
-        } else {
-            write!(f, "{}/{}", self.numer, self.denom)
+macro_rules! impl_formatting {
+    ($fmt_trait:ident, $prefix:expr, $fmt_str:expr, $fmt_alt:expr) => {
+        impl<T: $fmt_trait + Clone + Integer> $fmt_trait for Ratio<T> {
+            #[cfg(feature = "std")]
+            fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+                let pre_pad = if self.denom.is_one() {
+                    format!($fmt_str, self.numer)
+                } else {
+                    if f.alternate() {
+                        format!(concat!($fmt_str, "/", $fmt_alt), self.numer, self.denom)
+                    } else {
+                        format!(concat!($fmt_str, "/", $fmt_str), self.numer, self.denom)
+                    }
+                };
+                if f.sign_plus() {
+                    let pre_pad = pre_pad.trim_start_matches('-');
+                    let non_negative = self.numer >= T::zero();
+                    f.pad_integral(non_negative, $prefix, pre_pad)
+                } else {
+                    f.pad_integral(true, $prefix, &pre_pad)
```

Comment:
Always `true` doesn't work if there's padding on a negative number, e.g. `println!("{:05}", -1i8)` shows `-0001`, but yours will show `000-1`. Maybe this can strip the sign and use the same `non_negative` as above? Might not even need separate cases anymore.